Add single-process DDP accuracy support to dynamo benchmark suite

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88523
* #88521
* __->__ #88511
* #88480
* #88460
* #88435

- does not intend to support multi-process, as that is more complex
  and we have torchbench scripts for that
- currently only works in accuracy mode as this was the main goal,
  but could be extended for measuring single-gpu perf impact of
  graph breaks

Run with

`python benchmarks/dynamo/torchbench.py --inductor --training --accuracy --only hf_Bert --ddp`

Example output
```
cuda train hf_Bert
[2022-11-04 18:52:08,304] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to complex input striding
PASS
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

wconstab committed:
Add single-process DDP accuracy support to dynamo benchmark suite

- does not intend to support multi-process, as that is more complex
  and we have torchbench scripts for that
- currently only works in accuracy mode as this was the main goal,
  but could be extended for measuring single-gpu perf impact of
  graph breaks

Run with

`python benchmarks/dynamo/torchbench.py --inductor --training --accuracy --only hf_Bert --ddp`

Example output
```
cuda train hf_Bert
[2022-11-04 18:52:08,304] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to complex input striding
PASS
```

[ghstack-poisoned]

aazzolini commented on benchmarks/dynamo/common.py:
```diff
@@ -1557,6 +1586,10 @@ def run(runner, args, original_dir=None):
                 else CI_SKIP_INDCUTOR_INFERENCE
             )
+    if args.ddp:
+        assert args.accuracy, "TODO: hook DDP support to --speedup mode"
```
Can you provide a more instructive error message here? not clear, what does args.accuracy + arg.ddp have to do with --speedup mode?

aazzolini commented on benchmarks/dynamo/common.py:
```diff
@@ -786,6 +788,22 @@ def inner(*args, **kwargs):
+    try:
+        if should_init_distributed:
+            torch.cuda.set_device(rank)
+            os.environ['MASTER_ADDR'] = 'localhost'
+            os.environ['MASTER_PORT'] = '6789'
```
Will this break if 2 instances of this benchmark are running on the same machine at once?

wconstab replied:
yes. :(

is there a good pattern i can copy for getting a free port?

davidberard98 commented on benchmarks/dynamo/common.py:
```diff
@@ -1529,10 +1558,10 @@ def parse_args():
-        runner, args, original_dir
-    )
-
+    with maybe_init_distributed(args.ddp and args.only):
```
is `args.only` to avoid any distributed state issues?

wconstab replied:
no, its a pattern for this suite.  `--only` will be present for any process that is actually going to run a benchmark.
1) you ran with `--only` yourself: it literally starts the benchmark.  (and we want to init ddp stuff)
2) you are sweeping over many models: the outer script (invoked without only) will subprocess once per model, and add the `--only {name}` arg to the subprocess

davidberard98 reviewed (approved):
Looks good!

wconstab committed:
Update on "Add single-process DDP accuracy support to dynamo benchmark suite"


- does not intend to support multi-process, as that is more complex
  and we have torchbench scripts for that
- currently only works in accuracy mode as this was the main goal,
  but could be extended for measuring single-gpu perf impact of
  graph breaks

Run with

`python benchmarks/dynamo/torchbench.py --inductor --training --accuracy --only hf_Bert --ddp`

Example output
```
cuda train hf_Bert
[2022-11-04 18:52:08,304] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to complex input striding
PASS
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab committed:
Update on "Add single-process DDP accuracy support to dynamo benchmark suite"


- does not intend to support multi-process, as that is more complex
  and we have torchbench scripts for that
- currently only works in accuracy mode as this was the main goal,
  but could be extended for measuring single-gpu perf impact of
  graph breaks

Run with

`python benchmarks/dynamo/torchbench.py --inductor --training --accuracy --only hf_Bert --ddp`

Example output
```
cuda train hf_Bert
[2022-11-04 18:52:08,304] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to complex input striding
PASS
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab committed:
Update on "Add single-process DDP accuracy support to dynamo benchmark suite"


- does not intend to support multi-process, as that is more complex
  and we have torchbench scripts for that
- currently only works in accuracy mode as this was the main goal,
  but could be extended for measuring single-gpu perf impact of
  graph breaks

Run with

`python benchmarks/dynamo/torchbench.py --inductor --training --accuracy --only hf_Bert --ddp`

Example output
```
cuda train hf_Bert
[2022-11-04 18:52:08,304] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to complex input striding
PASS
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge -f "flaky CI"
